Extension is not working

**Describe the bug**
hey i've installed latest **deno version** **```1.3.0```** , and have vscode insiders installed , 
before one week everything was working fine , yesterday i opened vscode , but the **suggestions are not working**
i have vscode insiders version **```1.48.0```** , and whenever i try to import some urls it gives red line errors

**Expected behavior**

extension should give suggestions, 

**Versions**

vscode: 1.48.0
deno: 1.3.0
extension: 2.0.4

@lucacasonato Same Error here.
I also get the "Failed to load module 'typescript-deno-plugin' from" Error.

Environment:
Ubuntu 18
VSCode v1.48.0
I tried it with Deno version 1.2.3 and 1.3.0
denoland.vscode-deno Version 2.0.10

TS Server Log:
Info 25   [12:18:41.84] Plugin validation succeded
Info 26   [12:18:41.84] Loading global plugin typescript-deno-plugin
Info 27   [12:18:41.84] Enabling plugin typescript-deno-plugin from candidate paths: /usr/share/code/resources/app/extensions/typescript-language-features,/home/abc/.vscode/extensions/denoland.vscode-deno-2.0.10,/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js/../../..
Info 28   [12:18:41.85] Loading typescript-deno-plugin from /usr/share/code/resources/app/extensions/typescript-language-features (resolved to /usr/share/code/resources/app/extensions/typescript-language-features/node_modules)
Info 29   [12:18:41.135] Loading typescript-deno-plugin from /home/abc/.vscode/extensions/denoland.vscode-deno-2.0.10 (resolved to /home/abc/.vscode/extensions/denoland.vscode-deno-2.0.10/node_modules)
Info 30   [12:18:41.157] Loading typescript-deno-plugin from /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js/../../.. (resolved to /usr/share/code/resources/app/extensions/node_modules/node_modules)
Info 31   [12:18:41.160] Failed to load module 'typescript-deno-plugin' from /usr/share/code/resources/app/extensions/typescript-language-features/node_modules: Error: Could not resolve JS module 'typescript-deno-plugin' starting at '/usr/share/code/resources/app/extensions/typescript-language-features/node_modules'. Looked in: /usr/share/code/resources/app/extensions/typescript-language-features/node_modules/typescript-deno-plugin/package.json, /usr/share/code/resources/app/extensions/typescript-language-features/node_modules/typescript-deno-plugin.js, /usr/share/code/resources/app/extensions/typescript-language-features/node_modules/typescript-deno-plugin.jsx, /usr/share/code/resources/app/extensions/typescript-language-features/node_modules/typescript-deno-plugin/index.js, /usr/share/code/resources/app/extensions/typescript-language-features/node_modules/typescript-deno-plugin/index.jsx, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin/package.json, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin.js, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin.jsx, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin/index.js, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin/index.jsx, /usr/share/code/resources/app/node_modules/typescript-deno-plugin/package.json, /usr/share/code/resources/app/node_modules/typescript-deno-plugin.js, /usr/share/code/resources/app/node_modules/typescript-deno-plugin.jsx, /usr/share/code/resources/app/node_modules/typescript-deno-plugin/index.js, /usr/share/code/resources/app/node_modules/typescript-deno-plugin/index.jsx, /usr/share/code/resources/node_modules/typescript-deno-plugin/package.json, /usr/share/code/resources/node_modules/typescript-deno-plugin.js, /usr/share/code/resources/node_modules/typescript-deno-plugin.jsx, /usr/share/code/resources/node_modules/typescript-deno-plugin/index.js, /usr/share/code/resources/node_modules/typescript-deno-plugin/index.jsx, /usr/share/code/node_modules/typescript-deno-plugin/package.json, /usr/share/code/node_modules/typescript-deno-plugin.js, /usr/share/code/node_modules/typescript-deno-plugin.jsx, /usr/share/code/node_modules/typescript-deno-plugin/index.js, /usr/share/code/node_modules/typescript-deno-plugin/index.jsx, /usr/share/node_modules/typescript-deno-plugin/package.json, /usr/share/node_modules/typescript-deno-plugin.js, /usr/share/node_modules/typescript-deno-plugin.jsx, /usr/share/node_modules/typescript-deno-plugin/index.js, /usr/share/node_modules/typescript-deno-plugin/index.jsx, /usr/node_modules/typescript-deno-plugin/package.json, /usr/node_modules/typescript-deno-plugin.js, /usr/node_modules/typescript-deno-plugin.jsx, /usr/node_modules/typescript-deno-plugin/index.js, /usr/node_modules/typescript-deno-plugin/index.jsx, /node_modules/typescript-deno-plugin/package.json, /node_modules/typescript-deno-plugin.js, /node_modules/typescript-deno-plugin.jsx, /node_modules/typescript-deno-plugin/index.js, /node_modules/typescript-deno-plugin/index.jsx
    at Object.resolveJSModule (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:29954:19)
    at Object.sys.require (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:151195:45)
    at Function.Project.resolveModule (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:142740:35)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143761:36
    at Object.firstDefined (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:197:26)
    at InferredProject.Project.enablePlugin (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143760:41)
    at _loop_2 (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143744:32)
    at InferredProject.Project.enableGlobalPlugins (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143750:25)
    at new InferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143921:23)
    at ProjectService.createInferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145981:31)
    at ProjectService.getOrCreateInferredProjectForProjectRootPathIfEnabled (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145919:33)
    at ProjectService.assignOrphanScriptInfoToInferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145163:36)
    at ProjectService.assignProjectToOpenedScriptInfo (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146672:26)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146929:97
    at Object.flatMap (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:465:25)
    at ProjectService.applyChangesInOpenFiles (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146929:41)
    at Session.handlers.ts.createMapFromTemplate._a.<computed> (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:147792:46)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149555:88
    at IOSession.Session.executeWithRequestId (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149546:28)
    at IOSession.Session.executeCommand (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149555:33)
    at IOSession.Session.onMessage (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149579:35)
    at Interface.<anonymous> (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:150895:27)
    at Interface.emit (events.js:203:13)
    at Interface._onLine (readline.js:316:10)
    at Interface._normalWrite (readline.js:461:12)
    at Socket.ondata (readline.js:172:10)
    at Socket.emit (events.js:203:13)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:276:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:166:17)
Info 32   [12:18:41.160] Failed to load module 'typescript-deno-plugin' from /home/abc/.vscode/extensions/denoland.vscode-deno-2.0.10/node_modules: Error: Cannot find module 'typescript/lib/tsserverlibrary'
Require stack:
- /home/abc/.vscode/extensions/denoland.vscode-deno-2.0.10/node_modules/typescript-deno-plugin/out/deno_modules.js
- /home/abc/.vscode/extensions/denoland.vscode-deno-2.0.10/node_modules/typescript-deno-plugin/out/index.js
- /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:764:15)
    at Module._load (internal/modules/cjs/loader.js:669:27)
    at Module._load (electron/js2c/asar.js:717:26)
    at Function.Module._load (electron/js2c/asar.js:717:26)
    at Module.require (internal/modules/cjs/loader.js:822:19)
    at require (internal/modules/cjs/helpers.js:68:18)
    at Object.<anonymous> (/home/abc/.vscode/extensions/denoland.vscode-deno-2.0.10/node_modules/typescript-deno-plugin/out/deno_modules.js:7:43)
    at Module._compile (internal/modules/cjs/loader.js:927:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:939:10)
    at Module.load (internal/modules/cjs/loader.js:782:32)
    at Module._load (internal/modules/cjs/loader.js:695:12)
    at Module._load (electron/js2c/asar.js:717:26)
    at Function.Module._load (electron/js2c/asar.js:717:26)
    at Module.require (internal/modules/cjs/loader.js:822:19)
    at require (internal/modules/cjs/helpers.js:68:18)
    at Object.<anonymous> (/home/abc/.vscode/extensions/denoland.vscode-deno-2.0.10/node_modules/typescript-deno-plugin/out/index.js:16:24)
    at Module._compile (internal/modules/cjs/loader.js:927:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:939:10)
    at Module.load (internal/modules/cjs/loader.js:782:32)
    at Module._load (internal/modules/cjs/loader.js:695:12)
    at Module._load (electron/js2c/asar.js:717:26)
    at Function.Module._load (electron/js2c/asar.js:717:26)
    at Module.require (internal/modules/cjs/loader.js:822:19)
    at require (internal/modules/cjs/helpers.js:68:18)
    at Object.sys.require (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:151195:34)
    at Function.Project.resolveModule (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:142740:35)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143761:36
    at Object.firstDefined (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:197:26)
    at InferredProject.Project.enablePlugin (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143760:41)
    at _loop_2 (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143744:32)
    at InferredProject.Project.enableGlobalPlugins (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143750:25)
    at new InferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143921:23)
    at ProjectService.createInferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145981:31)
    at ProjectService.getOrCreateInferredProjectForProjectRootPathIfEnabled (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145919:33)
    at ProjectService.assignOrphanScriptInfoToInferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145163:36)
    at ProjectService.assignProjectToOpenedScriptInfo (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146672:26)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146929:97
    at Object.flatMap (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:465:25)
    at ProjectService.applyChangesInOpenFiles (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146929:41)
    at Session.handlers.ts.createMapFromTemplate._a.<computed> (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:147792:46)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149555:88
    at IOSession.Session.executeWithRequestId (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149546:28)
    at IOSession.Session.executeCommand (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149555:33)
    at IOSession.Session.onMessage (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149579:35)
    at Interface.<anonymous> (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:150895:27)
    at Interface.emit (events.js:203:13)
    at Interface._onLine (readline.js:316:10)
    at Interface._normalWrite (readline.js:461:12)
    at Socket.ondata (readline.js:172:10)
    at Socket.emit (events.js:203:13)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:276:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:166:17)
Info 33   [12:18:41.161] Failed to load module 'typescript-deno-plugin' from /usr/share/code/resources/app/extensions/node_modules/node_modules: Error: Could not resolve JS module 'typescript-deno-plugin' starting at '/usr/share/code/resources/app/extensions/node_modules/node_modules'. Looked in: /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin/package.json, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin.js, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin.jsx, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin/index.js, /usr/share/code/resources/app/extensions/node_modules/typescript-deno-plugin/index.jsx, /usr/share/code/resources/app/node_modules/typescript-deno-plugin/package.json, /usr/share/code/resources/app/node_modules/typescript-deno-plugin.js, /usr/share/code/resources/app/node_modules/typescript-deno-plugin.jsx, /usr/share/code/resources/app/node_modules/typescript-deno-plugin/index.js, /usr/share/code/resources/app/node_modules/typescript-deno-plugin/index.jsx, /usr/share/code/resources/node_modules/typescript-deno-plugin/package.json, /usr/share/code/resources/node_modules/typescript-deno-plugin.js, /usr/share/code/resources/node_modules/typescript-deno-plugin.jsx, /usr/share/code/resources/node_modules/typescript-deno-plugin/index.js, /usr/share/code/resources/node_modules/typescript-deno-plugin/index.jsx, /usr/share/code/node_modules/typescript-deno-plugin/package.json, /usr/share/code/node_modules/typescript-deno-plugin.js, /usr/share/code/node_modules/typescript-deno-plugin.jsx, /usr/share/code/node_modules/typescript-deno-plugin/index.js, /usr/share/code/node_modules/typescript-deno-plugin/index.jsx, /usr/share/node_modules/typescript-deno-plugin/package.json, /usr/share/node_modules/typescript-deno-plugin.js, /usr/share/node_modules/typescript-deno-plugin.jsx, /usr/share/node_modules/typescript-deno-plugin/index.js, /usr/share/node_modules/typescript-deno-plugin/index.jsx, /usr/node_modules/typescript-deno-plugin/package.json, /usr/node_modules/typescript-deno-plugin.js, /usr/node_modules/typescript-deno-plugin.jsx, /usr/node_modules/typescript-deno-plugin/index.js, /usr/node_modules/typescript-deno-plugin/index.jsx, /node_modules/typescript-deno-plugin/package.json, /node_modules/typescript-deno-plugin.js, /node_modules/typescript-deno-plugin.jsx, /node_modules/typescript-deno-plugin/index.js, /node_modules/typescript-deno-plugin/index.jsx
    at Object.resolveJSModule (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:29954:19)
    at Object.sys.require (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:151195:45)
    at Function.Project.resolveModule (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:142740:35)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143761:36
    at Object.firstDefined (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:197:26)
    at InferredProject.Project.enablePlugin (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143760:41)
    at _loop_2 (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143744:32)
    at InferredProject.Project.enableGlobalPlugins (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143750:25)
    at new InferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:143921:23)
    at ProjectService.createInferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145981:31)
    at ProjectService.getOrCreateInferredProjectForProjectRootPathIfEnabled (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145919:33)
    at ProjectService.assignOrphanScriptInfoToInferredProject (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:145163:36)
    at ProjectService.assignProjectToOpenedScriptInfo (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146672:26)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146929:97
    at Object.flatMap (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:465:25)
    at ProjectService.applyChangesInOpenFiles (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:146929:41)
    at Session.handlers.ts.createMapFromTemplate._a.<computed> (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:147792:46)
    at /usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149555:88
    at IOSession.Session.executeWithRequestId (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149546:28)
    at IOSession.Session.executeCommand (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149555:33)
    at IOSession.Session.onMessage (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:149579:35)
    at Interface.<anonymous> (/usr/share/code/resources/app/extensions/node_modules/typescript/lib/tsserver.js:150895:27)
    at Interface.emit (events.js:203:13)
    at Interface._onLine (readline.js:316:10)
    at Interface._normalWrite (readline.js:461:12)
    at Socket.ondata (readline.js:172:10)
    at Socket.emit (events.js:203:13)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:276:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:166:17)
Info 34   [12:18:41.161] Couldn't find typescript-deno-plugin

--- a/package.json
+++ b/package.json
@@ -2,7 +2,7 @@
   "name": "vscode-deno",
   "displayName": "Deno",
   "description": "Deno support for VSCode",
-  "version": "2.0.10",
+  "version": "2.0.11",
   "publisher": "denoland",
   "icon": "deno.png",
   "galleryBanner": {
@@ -199,7 +199,7 @@
     "deep-equal": "2.0.3",
     "deepmerge": "^4.2.2",
     "json5": "^2.1.1",
-    "typescript-deno-plugin": "1.31.0",
+    "typescript-deno-plugin": "./typescript-deno-plugin",
     "semver": "7.3.2"
   }
 }

--- a/typescript-deno-plugin/.npmignore
+++ b/typescript-deno-plugin/.npmignore
@@ -1,0 +1,1 @@
+tsconfig.json

--- a/typescript-deno-plugin/LICENSE
+++ b/typescript-deno-plugin/LICENSE
@@ -1,0 +1,21 @@
+MIT License
+
+Copyright (c) 2019 justjavac
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.

--- a/typescript-deno-plugin/package.json
+++ b/typescript-deno-plugin/package.json
@@ -1,0 +1,16 @@
+{
+  "name": "typescript-deno-plugin",
+  "version": "0.1.0",
+  "description": "Deno language service plugin for TypeScript",
+  "main": "./out/typescript-deno-plugin/src/index.js",
+  "dependencies": {
+    "deep-equal": "^2.0.1",
+    "deepmerge": "^4.2.2",
+    "json5": "^2.1.1"
+  },
+  "devDependencies": {
+    "@types/deep-equal": "1.0.1",
+    "@types/json5": "0.0.30",
+    "typescript": "3.8.3"
+  }
+}

--- a/typescript-deno-plugin/src/index.ts
+++ b/typescript-deno-plugin/src/index.ts
@@ -1,0 +1,12 @@
+import ts_module from "typescript/lib/tsserverlibrary";
+import { DenoPlugin } from "./plugin";
+
+module.exports = function init({
+  typescript,
+}: {
+  typescript: typeof ts_module;
+}) {
+  const plugin = new DenoPlugin(typescript);
+
+  return plugin;
+};

--- a/typescript-deno-plugin/src/logger.ts
+++ b/typescript-deno-plugin/src/logger.ts
@@ -1,0 +1,19 @@
+import ts_module from "typescript/lib/tsserverlibrary";
+
+export class Logger {
+  public static forPlugin(
+    name: string,
+    info: ts_module.server.PluginCreateInfo
+  ) {
+    return new Logger(name, info.project.projectService.logger);
+  }
+
+  private constructor(
+    private readonly name: string,
+    private readonly logger: ts_module.server.Logger
+  ) {}
+
+  public info(message: string) {
+    this.logger.info(`[${this.name}] ${JSON.stringify(message)}`);
+  }
+}

--- a/typescript-deno-plugin/src/plugin.ts
+++ b/typescript-deno-plugin/src/plugin.ts
@@ -1,0 +1,515 @@
+import * as path from "path";
+
+import merge from "deepmerge";
+import ts_module from "typescript/lib/tsserverlibrary";
+
+import { Logger } from "./logger";
+import { Configuration, ConfigurationField } from "../../core/configuration";
+import { getDenoDts } from "../../core/deno";
+import { ModuleResolver } from "../../core/module_resolver";
+import { CacheModule } from "../../core/deno_cache";
+import { normalizeFilepath, isUntitledDocument } from "../../core/util";
+import { normalizeImportStatement } from "../../core/deno_normalize_import_statement";
+import { getImportModules } from "../../core/deno_deps";
+
+const ignoredCompilerOptions: readonly string[] = [
+  "allowSyntheticDefaultImports",
+  "baseUrl",
+  "build",
+  "composite",
+  "declaration",
+  "declarationDir",
+  "declarationMap",
+  "diagnostics",
+  "downlevelIteration",
+  "emitBOM",
+  "emitDeclarationOnly",
+  "esModuleInterop",
+  "extendedDiagnostics",
+  "forceConsistentCasingInFileNames",
+  "help",
+  "importHelpers",
+  "incremental",
+  "inlineSourceMap",
+  "inlineSources",
+  "init",
+  "isolatedModules",
+  "listEmittedFiles",
+  "listFiles",
+  "mapRoot",
+  "maxNodeModuleJsDepth",
+  "module",
+  "moduleResolution",
+  "newLine",
+  "noEmit",
+  "noEmitHelpers",
+  "noEmitOnError",
+  "noLib",
+  "noResolve",
+  "out",
+  "outDir",
+  "outFile",
+  "paths",
+  "preserveSymlinks",
+  "preserveWatchOutput",
+  "pretty",
+  "rootDir",
+  "rootDirs",
+  "showConfig",
+  "skipDefaultLibCheck",
+  "skipLibCheck",
+  "sourceMap",
+  "sourceRoot",
+  "stripInternal",
+  "target",
+  "traceResolution",
+  "tsBuildInfoFile",
+  "types",
+  "typeRoots",
+  "version",
+  "watch",
+];
+
+export class DenoPlugin implements ts_module.server.PluginModule {
+  // plugin name
+  static readonly PLUGIN_NAME = "typescript-deno-plugin";
+  private readonly configurationManager = new Configuration();
+  // see https://github.com/denoland/deno/blob/2debbdacb935cfe1eb7bb8d1f40a5063b339d90b/js/compiler.ts#L159-L170
+  private readonly DEFAULT_OPTIONS: ts_module.CompilerOptions = {
+    allowJs: true,
+    checkJs: false,
+    strict: true,
+    esModuleInterop: true,
+    jsx: this.typescript.JsxEmit.React,
+    module: this.typescript.ModuleKind.ESNext,
+    moduleResolution: this.typescript.ModuleResolutionKind.NodeJs,
+    outDir: "$deno$",
+    resolveJsonModule: true,
+    sourceMap: true,
+    stripComments: true,
+    target: this.typescript.ScriptTarget.ESNext,
+    noEmit: true,
+    noEmitHelpers: true,
+  };
+  // No matter how tsconfig.json is set in the working directory
+  // It will always overwrite the configuration
+  private readonly MUST_OVERWRITE_OPTIONS: ts_module.CompilerOptions = {
+    jsx: this.DEFAULT_OPTIONS.jsx,
+    module: this.DEFAULT_OPTIONS.module,
+    moduleResolution: this.DEFAULT_OPTIONS.moduleResolution,
+    resolveJsonModule: this.DEFAULT_OPTIONS.resolveJsonModule,
+    strict: this.DEFAULT_OPTIONS.strict,
+    noEmit: this.DEFAULT_OPTIONS.noEmit,
+    noEmitHelpers: this.DEFAULT_OPTIONS.noEmitHelpers,
+    target: this.typescript.ScriptTarget.ESNext,
+  };
+  private logger!: Logger;
+
+  constructor(private readonly typescript: typeof ts_module) {}
+
+  create(info: ts_module.server.PluginCreateInfo): ts_module.LanguageService {
+    const { project, languageService, languageServiceHost } = info;
+    const projectDirectory = project.getCurrentDirectory();
+
+    function getRealPath(filepath: string): string {
+      return project.realpath ? project.realpath(filepath) : filepath;
+    }
+
+    // TypeScript plugins have a `cwd` of `/`, which causes issues with import resolution.
+    process.chdir(projectDirectory);
+
+    this.logger = Logger.forPlugin(DenoPlugin.PLUGIN_NAME, info);
+
+    this.logger.info(`Create typescript-deno-plugin`);
+    const getCompilationSettings = languageServiceHost.getCompilationSettings.bind(
+      languageServiceHost
+    );
+    const getScriptFileNames = languageServiceHost.getScriptFileNames.bind(
+      languageServiceHost
+    );
+    // ref https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#customizing-module-resolution
+    const resolveModuleNames = languageServiceHost.resolveModuleNames?.bind(
+      languageServiceHost
+    );
+    const getSemanticDiagnostics = languageService.getSemanticDiagnostics.bind(
+      languageService
+    );
+
+    const resolveTypeReferenceDirectives = languageServiceHost.resolveTypeReferenceDirectives?.bind(
+      languageServiceHost
+    );
+
+    const getCompletionEntryDetails = languageService.getCompletionEntryDetails.bind(
+      languageService
+    );
+
+    languageServiceHost.getCompilationSettings = () => {
+      const projectConfig = getCompilationSettings();
+
+      if (!this.configurationManager.config.enable) {
+        return projectConfig;
+      }
+
+      // delete the option which ignore by Deno
+      // see https://github.com/denoland/deno/blob/bced52505f/cli/js/compiler/host.ts#L65-L121
+      for (const option in projectConfig) {
+        if (ignoredCompilerOptions.includes(option)) {
+          delete projectConfig[option];
+        }
+      }
+
+      const compilationSettings = merge(
+        merge(this.DEFAULT_OPTIONS, projectConfig),
+        this.MUST_OVERWRITE_OPTIONS
+      );
+
+      return compilationSettings;
+    };
+
+    languageServiceHost.getScriptFileNames = () => {
+      const scriptFileNames = getScriptFileNames();
+
+      if (!this.configurationManager.config.enable) {
+        return scriptFileNames;
+      }
+
+      // Get typescript declaration File
+      const dtsFiles = [
+        getDenoDts(!!this.configurationManager.config.unstable),
+      ];
+
+      const iterator = new Set(dtsFiles).entries();
+
+      for (const [, filepath] of iterator) {
+        scriptFileNames.push(filepath);
+      }
+
+      return scriptFileNames;
+    };
+
+    if (resolveTypeReferenceDirectives) {
+      languageServiceHost.resolveTypeReferenceDirectives = (
+        typeDirectiveNames: string[],
+        containingFile: string,
+        ...rest
+      ) => {
+        if (!this.configurationManager.config.enable) {
+          return resolveTypeReferenceDirectives(
+            typeDirectiveNames,
+            containingFile,
+            ...rest
+          );
+        }
+
+        // containingFile may be `untitled: ^ Untitled-1`
+        const realContainingFile = isUntitledDocument(containingFile)
+          ? path.join(project.getCurrentDirectory(), "untitled")
+          : // in Windows.
+            // containingFile may be a unix-like style
+            // eg. c:/Users/admin/path/to/file.ts
+            // This is not a legal file path in Windows
+            // It will cause a series of bugs, so here we get the real file path
+            getRealPath(containingFile);
+
+        if (!this.typescript.sys.fileExists(realContainingFile)) {
+          return [];
+        }
+
+        const importMapsFilepath = this.configurationManager.config.import_map
+          ? path.isAbsolute(this.configurationManager.config.import_map)
+            ? this.configurationManager.config.import_map
+            : path.resolve(
+                project.getCurrentDirectory(),
+                this.configurationManager.config.import_map
+              )
+          : undefined;
+
+        const resolver = ModuleResolver.create(
+          realContainingFile,
+          importMapsFilepath
+        );
+
+        const result: (
+          | ts_module.ResolvedTypeReferenceDirective
+          | undefined
+        )[] = [];
+
+        for (const typeDirectiveName of typeDirectiveNames) {
+          const [resolvedModule] = resolver.resolveModules([typeDirectiveName]);
+
+          // if module found. then return the module file
+          if (resolvedModule) {
+            const target: ts_module.ResolvedTypeReferenceDirective = {
+              primary: false,
+              resolvedFileName: resolvedModule.filepath,
+            };
+
+            result.push(target);
+            continue;
+          }
+
+          // If the module does not exist, then apply the native reference method
+          const [target] = resolveTypeReferenceDirectives(
+            [typeDirectiveName],
+            containingFile,
+            ...rest
+          );
+
+          result.push(target);
+        }
+
+        return result;
+      };
+    }
+
+    languageService.getSemanticDiagnostics = (filename: string) => {
+      const diagnostics = getSemanticDiagnostics(filename);
+
+      if (!this.configurationManager.config.enable) {
+        return diagnostics;
+      }
+
+      // ref: https://github.com/denoland/deno/blob/da8cb408c878aa6e90542e26173f1f14b5254d29/cli/js/compiler/util.ts#L262
+      const ignoredDiagnostics = [
+        // TS2306: File 'file:///Users/rld/src/deno/cli/tests/subdir/amd_like.js' is
+        // not a module.
+        2306,
+        // TS1375: 'await' expressions are only allowed at the top level of a file
+        // when that file is a module, but this file has no imports or exports.
+        // Consider adding an empty 'export {}' to make this file a module.
+        1375,
+        // TS1103: 'for-await-of' statement is only allowed within an async function
+        // or async generator.
+        1103,
+        // TS2691: An import path cannot end with a '.ts' extension. Consider
+        // importing 'bad-module' instead.
+        2691,
+        // TS5009: Cannot find the common subdirectory path for the input files.
+        5009,
+        // TS5055: Cannot write file
+        // 'http://localhost:4545/cli/tests/subdir/mt_application_x_javascript.j4.js'
+        // because it would overwrite input file.
+        5055,
+        // TypeScript is overly opinionated that only CommonJS modules kinds can
+        // support JSON imports.  Allegedly this was fixed in
+        // Microsoft/TypeScript#26825 but that doesn't seem to be working here,
+        // so we will ignore complaints about this compiler setting.
+        5070,
+        // TS7016: Could not find a declaration file for module '...'. '...'
+        // implicitly has an 'any' type.  This is due to `allowJs` being off by
+        // default but importing of a JavaScript module.
+        7016,
+      ];
+
+      return diagnostics.filter((v) => !ignoredDiagnostics.includes(v.code));
+    };
+
+    languageService.getCompletionEntryDetails = (
+      fileName: string,
+      position: number,
+      name: string,
+      formatOptions?:
+        | ts_module.FormatCodeOptions
+        | ts_module.FormatCodeSettings,
+      source?: string,
+      preferences?: ts_module.UserPreferences
+    ) => {
+      const details = getCompletionEntryDetails(
+        fileName,
+        position,
+        name,
+        formatOptions,
+        source,
+        preferences
+      );
+
+      if (!this.configurationManager.config.enable) {
+        return details;
+      }
+
+      if (details) {
+        // modifiers maybe contain multiple values. eg `export,declare`
+        const modifiers = details.kindModifiers.split(",") || [];
+
+        if (
+          modifiers.includes("export") &&
+          details.codeActions &&
+          details.codeActions.length
+        ) {
+          for (const ca of details.codeActions) {
+            for (const change of ca.changes) {
+              if (!change.isNewFile) {
+                for (const tc of change.textChanges) {
+                  tc.newText = normalizeImportStatement(
+                    fileName,
+                    tc.newText,
+                    this.logger
+                  );
+                }
+              }
+            }
+          }
+        }
+
+        if (details.source && details.source.length) {
+          for (const source of details.source) {
+            if (source.kind === "text") {
+              // text is always unix style
+              const text = source.text;
+
+              const absoluteFilepath = path.resolve(
+                normalizeFilepath(path.dirname(fileName)),
+                normalizeFilepath(text)
+              );
+
+              if (path.isAbsolute(absoluteFilepath)) {
+                const denoCache = CacheModule.create(
+                  absoluteFilepath,
+                  this.logger
+                );
+
+                if (denoCache) {
+                  source.text = denoCache.meta.url.href;
+                }
+              }
+            }
+          }
+        }
+      }
+
+      return details;
+    };
+
+    if (resolveModuleNames) {
+      languageServiceHost.resolveModuleNames = (
+        moduleNames: string[],
+        containingFile: string,
+        ...rest
+      ): (
+        | ts_module.ResolvedModule
+        | ts_module.ResolvedModuleFull
+        | undefined
+      )[] => {
+        if (!this.configurationManager.config.enable) {
+          return resolveModuleNames(moduleNames, containingFile, ...rest);
+        }
+
+        // containingFile may be `untitled: ^ Untitled-1`
+        const realContainingFile = isUntitledDocument(containingFile)
+          ? path.join(project.getCurrentDirectory(), "untitled")
+          : // in Windows.
+            // containingFile may be a unix-like style
+            // eg. c:/Users/admin/path/to/file.ts
+            // This is not a legal file path in Windows
+            // It will cause a series of bugs, so here we get the real file path
+            getRealPath(containingFile);
+
+        const importMapsFilepath = this.configurationManager.config.import_map
+          ? path.isAbsolute(this.configurationManager.config.import_map)
+            ? this.configurationManager.config.import_map
+            : path.resolve(
+                project.getCurrentDirectory(),
+                this.configurationManager.config.import_map
+              )
+          : undefined;
+
+        const resolver = ModuleResolver.create(
+          realContainingFile,
+          importMapsFilepath
+        );
+
+        const content = this.typescript.sys.readFile(containingFile, "utf8");
+
+        // handle @deno-types
+        if (content && content.indexOf("// @deno-types=") >= 0) {
+          const sourceFile = this.typescript.createSourceFile(
+            containingFile,
+            content,
+            this.typescript.ScriptTarget.ESNext,
+            true
+          );
+
+          const modules = getImportModules(this.typescript)(sourceFile);
+
+          for (const m of modules) {
+            if (m.hint) {
+              const index = moduleNames.findIndex((v) => v === m.moduleName);
+
+              moduleNames[index] = m.hint.text;
+            }
+          }
+        }
+
+        const resolvedModules = resolver.resolveModules(moduleNames);
+
+        // try resolve typeReferenceDirectives
+        for (const resolvedModule of resolvedModules) {
+          if (!resolvedModule) {
+            continue;
+          }
+
+          const content = this.typescript.sys.readFile(resolvedModule.filepath);
+
+          if (!content) {
+            continue;
+          }
+
+          const { typeReferenceDirectives } = this.typescript.preProcessFile(
+            content,
+            true,
+            true
+          );
+
+          if (!typeReferenceDirectives.length) {
+            continue;
+          }
+
+          const _resolver = ModuleResolver.create(
+            resolvedModule.filepath,
+            importMapsFilepath
+          );
+
+          const modules = _resolver.resolveModules(
+            typeReferenceDirectives.map((v) => v.fileName)
+          );
+
+          for (const m of modules) {
+            if (m) {
+              resolvedModule.origin = m.origin;
+              resolvedModule.filepath = m.filepath;
+            }
+          }
+        }
+
+        return resolvedModules.map((v) => {
+          if (!v) {
+            return v;
+          }
+
+          const result: ts_module.ResolvedModuleFull = {
+            extension: v.extension as ts_module.Extension,
+            isExternalLibraryImport: false,
+            resolvedFileName: v.filepath,
+          };
+
+          return result;
+        });
+      };
+    }
+
+    this.configurationManager.resolveFromVscode(projectDirectory);
+
+    this.configurationManager.onUpdatedConfig(() => {
+      project.refreshDiagnostics();
+      project.updateGraph();
+      languageService.getProgram()?.emit();
+    });
+
+    return languageService;
+  }
+
+  onConfigurationChanged(c: ConfigurationField) {
+    this.logger.info(`onConfigurationChanged: ${JSON.stringify(c)}`);
+    this.configurationManager.update(c);
+  }
+}

--- a/typescript-deno-plugin/tsconfig.json
+++ b/typescript-deno-plugin/tsconfig.json
@@ -1,0 +1,8 @@
+{
+  "extends": "../tsconfig.base.json",
+  "compilerOptions": {
+    "baseUrl": "./src",
+    "outDir": "../node_modules/typescript-deno-plugin/out"
+  },
+  "include": ["./src", "../core"]
+}

